Adding caching and merging of configurations

thorhs opened the pull request:
Configurations are now cached from each provider separately so that
we can merge them together when one provider has changed config.

The Web module also returns full config info through the HTML call,
but REST API is working on a separate web configuration that is
sent in just like any other.

EmileVauge commented:
Hi, thank you for your PR :) Awsome! I think it needs a little bit more work. See my comments :)

EmileVauge commented on web.go:
```diff
@@ -10,6 +10,10 @@ import (
 )
 
+var (
+	webConfiguration *Configuration
```
This variable is never initialized and therefore, the rest api doesn't work.

thorhs replied:
Yeah, I'm not using the API so I didn't test it, sorry for that.

The reason I split the API (webConfiguration) out  was because other providers, like marathon, will overwrite any changes to the marathon config on next update.If the API has changed any settings in marathon managed config, they will be overwritten.

One option would be to split the API in two segments, one for informational purposes that would represent the whole config, and one segment that would allow updates to a subset of the config (API managed) .  I'm not sufficiently familiar with how traefik is being used today by others so I can't tell if that would be a major change or not.

Another option would be to change the API to reflect the different providers, something like /api/{provider}/frontends|backends instead of /api/frontends.  /api/frontends could then be an informational API that would return the merged config.

Btw, should I add a new commit, or ammend this one?

EmileVauge commented:
> The reason I split the API (webConfiguration) out was because other providers, like marathon, will overwrite any changes to the marathon config on next update.If the API has changed any settings in marathon managed config, they will be overwritten.

Yep, I totally agree with you. We have to manage that point.

> Another option would be to change the API to reflect the different providers, something like /api/{provider}/frontends|backends instead of /api/frontends. /api/frontends could then be an informational API that would return the merged config.

I find this solution much nicer, even if it breaks the API.
I would also add an alias ```all``` that would returns every provider's config: ```/api/all/frontends|backends```, only for ```GET``` requests.
But I think it should be done in a separate PR.

@vdemeester what are your thoughts on that?

vdemeester commented:
@EmileVauge on the API side, I do agree that `/api/{provider}/` is definitely nicer, and I think we shouldn't mind breaking the API too much for now.

@thorhs let's not worry for commits for now :wink:. When we agree on the design and the code, we can easily squash or edit commits so do not worry about adding commits for now :smile_cat:.

thorhs commented:
I added a commit to initialize webConfiguration when we PUT /api, which seems to be what was done before.

EmileVauge commented:
@thorhs are you planning to do the API refactoring in this PR?

thorhs commented:
No, I think that is best done in another PR. 

Sent from my iPhone

> On 26 Sep 2015, at 10:26, Emile Vauge <notifications@github.com> wrote:
> 
> @thorhs are you planning to do the API refactoring in this PR?
> 
> —
> Reply to this email directly or view it on GitHub.
>

thorhs commented:
I decided to do the API refactoring anyway, it was simpler than I thought.

vdemeester commented:
@thorhs cool :) You need to rebase against the master though :sweat_smile: :blush:

thorhs commented:
Rebased

EmileVauge commented on web.go:
```diff
@@ -74,13 +82,14 @@ func GetHealthHandler(rw http.ResponseWriter, r *http.Request) {
 func GetBackendsHandler(rw http.ResponseWriter, r *http.Request) {
-	templatesRenderer.JSON(rw, http.StatusOK, currentConfiguration.Backends)
+	vars := mux.Vars(r)
+	templatesRenderer.JSON(rw, http.StatusOK, currentConfigurations[vars["provider"]].Backends)
```
You should check if currentConfigurations[vars["provider"]] is not nil

EmileVauge commented on web.go:
```diff
@@ -74,13 +82,14 @@ func GetHealthHandler(rw http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	id := vars["backend"]
-	if backend, ok := currentConfiguration.Backends[id]; ok {
+	if backend, ok := currentConfigurations[vars["provider"]].Backends[id]; ok {
```
You should check if currentConfigurations[vars["provider"]] is not nil

EmileVauge commented on web.go:
```diff
@@ -88,13 +97,14 @@ func GetBackendHandler(rw http.ResponseWriter, r *http.Request) {
 func GetFrontendsHandler(rw http.ResponseWriter, r *http.Request) {
-	templatesRenderer.JSON(rw, http.StatusOK, currentConfiguration.Frontends)
+	vars := mux.Vars(r)
+	templatesRenderer.JSON(rw, http.StatusOK, currentConfigurations[vars["provider"]].Frontends)
```
You should check if currentConfigurations[vars["provider"]] is not nil

EmileVauge commented on web.go:
```diff
@@ -88,13 +97,14 @@ func GetBackendHandler(rw http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	id := vars["frontend"]
-	if frontend, ok := currentConfiguration.Frontends[id]; ok {
+	if frontend, ok := currentConfigurations[vars["provider"]].Frontends[id]; ok {
```
You should check if currentConfigurations[vars["provider"]] is not nil

EmileVauge commented on web.go:
```diff
@@ -104,7 +114,7 @@ func GetFrontendHandler(rw http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	backend := vars["backend"]
-	if backend, ok := currentConfiguration.Backends[backend]; ok {
+	if backend, ok := currentConfigurations[vars["provider"]].Backends[backend]; ok {
```
You should check if currentConfigurations[vars["provider"]] is not nil

EmileVauge commented on web.go:
```diff
@@ -115,7 +125,7 @@ func GetServerHandler(rw http.ResponseWriter, r *http.Request) {
 	backend := vars["backend"]
 	server := vars["server"]
-	if backend, ok := currentConfiguration.Backends[backend]; ok {
+	if backend, ok := currentConfigurations[vars["provider"]].Backends[backend]; ok {
```
You should check if currentConfigurations[vars["provider"]] is not nil

EmileVauge commented:
Thank you :) Just minor comments on the api. Also, could you update the documentation on that new API? I will merge your PR directly when done :)

EmileVauge commented on traefik.go:
```diff
@@ -79,17 +86,25 @@ func main() {
+			log.Infof("Configuration receveived from provider %v: %+v", configMsg.providerName, configMsg.configuration)
+			if configMsg.configuration == nil {
 				log.Info("Skipping empty configuration")
-			} else if reflect.DeepEqual(currentConfiguration, configuration) {
```
I'm not sure about that... If we have an first conf update from file provider and then a second from docker provider, this test will compare the previous file config with the new docker one.

thorhs replied:
I think you may be looking at an old commit, currentConfiguration is not present in the latest push.  New line is: 
```go
    } else if reflect.DeepEqual(currentConfigurations[configMsg.providerName], configMsg.configuration) {
```

EmileVauge replied:
I need to sleep :tired_face:

EmileVauge commented on traefik.go:
```diff
@@ -79,17 +86,25 @@ func main() {
+
+				newConfigurationRouter, err := LoadConfig(newConfigurations, globalConfiguration)
 				if err == nil {
-					currentConfiguration = configuration
```
Still necessary?

thorhs replied:
This is so we don't try to enable a config that is invalid.  That is, if someone sends in a config that fails to 'load', it will get ignored and not activated.  We first copy the currentConfigurations to a new map, add the incoming config to the new map and try to LoadConfig it.  If that fails, we don't save the incoming config change and all is still good in the world.

EmileVauge replied:
Yep OK :)

EmileVauge commented:
OK @thorhs :), thank you and sorry for bothering again, there is still one last problem with https://github.com/EmileVauge/traefik/pull/12#discussion_r40508857. Also, could you update the documentation on the new API?

thorhs commented:
Hope these changes are sufficient.  I changed the doc to include {provider} and added error checks to the REST GET calls.

EmileVauge commented on web.go:
```diff
@@ -74,54 +82,80 @@ func GetHealthHandler(rw http.ResponseWriter, r *http.Request) {
+	serverId := vars["server"]
+	fmt.Printf("%v %v %v\n", providerId, backendId, serverId)
+	if provider, ok := currentConfigurations[providerId]; ok {
+		fmt.Printf("provider OK\n")
```
Ops ^^

EmileVauge commented on web.go:
```diff
@@ -74,54 +82,80 @@ func GetHealthHandler(rw http.ResponseWriter, r *http.Request) {
+	if provider, ok := currentConfigurations[providerId]; ok {
+		fmt.Printf("provider OK\n")
+		if backend, ok := provider.Backends[backendId]; ok {
+			fmt.Printf("backend OK\n")
```
Ops again

EmileVauge commented on web.go:
```diff
@@ -74,54 +82,80 @@ func GetHealthHandler(rw http.ResponseWriter, r *http.Request) {
+		if backend, ok := provider.Backends[backendId]; ok {
+			fmt.Printf("backend OK\n")
+			if server, ok := backend.Servers[serverId]; ok {
+				fmt.Printf("server OK\n")
```
Ops finally

EmileVauge commented:
Hi @thorhs could you just remove the 3 printf? Other than that LGTM for me. @vdemeester ?
I will rebase it tomorrow because the master has been moving quite ^^

vdemeester commented:
Once the comments from @EmileVauge are taken care of, LGTM too :wink:.
Thanks a lot, and sorry for the multiple rebase :stuck_out_tongue_closed_eyes:.

thorhs commented:
I have removed the Pritntf statements and rebased on to master.  Hope all the new stuff still works :)

I have compiled and done basic testing of marathon+file combo on my end, looks good.

vdemeester commented:
@thorhs yay :tada: Thanks again for your patience !

EmileVauge commented:
Thank you @thorhs ! Great :)

thorhs commented:
No problem, this directly benefits me so sharing it back is a no-brainer :)

On Tue, Sep 29, 2015 at 10:03 AM, Emile Vauge <notifications@github.com>
wrote:

> Thank you @thorhs <https://github.com/thorhs> ! Great :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/EmileVauge/traefik/pull/12#issuecomment-144012188>.
>
